File: algorithms/124.Binary_Tree_Maximum_Path_Sum/Binary_Tree_Maximum_Path_Sum.cpp
#include <bits/stdc++.h>

using namespace std;

// Definition for binary tree
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    int ret;
    int dfs(TreeNode *root) {
        if (root == NULL)
            return 0;
        int left_sum = max(0, dfs(root->left));
        int right_sum = max(0, dfs(root->right));
        ret = max(ret, left_sum + right_sum + root->val);
        return max(left_sum, right_sum) + root->val;
    }

public:
    int maxPathSum(TreeNode *root) {
        ret = INT_MIN;
        dfs(root);
        return ret;
    }
};
// LeetCode, Binary Tree Maximum Path Sum
// 时间复杂度O(n)，空间复杂度O(logn)
class Solution {
public:
    int maxPathSum(TreeNode *root) {
        max_sum = INT_MIN;
        dfs(root);
        return max_sum;
    }
private:
    int max_sum;
    int dfs(const TreeNode *root) {
        if (root == nullptr) return 0;
        int l = dfs(root->left);
        int r = dfs(root->right);
        int sum = root->val;
        if (l > 0) sum += l;
        if (r > 0) sum += r;
        max_sum = max(max_sum, sum);
        return max(r, l) > 0 ? max(r, l) + root->val : root->val;
    }
};

int main() {

    return 0;
}

File: algorithms/124.Binary_Tree_Maximum_Path_Sum/Binary_Tree_Maximum_Path_Sum.py
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def dfs(self, root):
        if root is None:
            return 0
        left = max(0, self.dfs(root.left))
        right = max(0, self.dfs(root.right))
        self.ret = max(self.ret, left + right + root.val)
        return max(left, right) + root.val

    # @param root, a tree node
    # @return an integer
    def maxPathSum(self, root):
        self.ret = -(1<<31)

        self.dfs(root)
        return self.ret

if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    result = Solution().maxPathSum(root)

